File: openai/typescript/src/services/assistants/runs.ts
import type { RequestOptions, PaginatedResponse } from '../../types/common.js';
import type { AssistantTool, AssistantToolResources } from './types.js';

export type RunStatus =
  | 'queued'
  | 'in_progress'
  | 'requires_action'
  | 'cancelling'
  | 'cancelled'
  | 'failed'
  | 'completed'
  | 'expired';

export interface Run {
  id: string;
  object: 'thread.run';
  created_at: number;
  thread_id: string;
  assistant_id: string;
  status: RunStatus;
  required_action: RunRequiredAction | null;
  last_error: RunError | null;
  expires_at: number | null;
  started_at: number | null;
  cancelled_at: number | null;
  failed_at: number | null;
  completed_at: number | null;
  model: string;
  instructions: string | null;
  tools: AssistantTool[];
  metadata: Record<string, string>;
  usage: RunUsage | null;
  temperature?: number | null;
  top_p?: number | null;
  max_prompt_tokens?: number | null;
  max_completion_tokens?: number | null;
  truncation_strategy?: RunTruncationStrategy;
  tool_choice?: 'none' | 'auto' | 'required' | { type: 'function'; function: { name: string } };
  response_format?: 'auto' | { type: 'text' | 'json_object' };
}

export interface RunRequiredAction {
  type: 'submit_tool_outputs';
  submit_tool_outputs: {
    tool_calls: RunToolCall[];
  };
}

export interface RunToolCall {
  id: string;
  type: 'function';
  function: {
    name: string;
    arguments: string;
  };
}

export interface RunError {
  code: string;
  message: string;
}

export interface RunUsage {
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
}

export interface RunTruncationStrategy {
  type: 'auto' | 'last_messages';
  last_messages?: number;
}

export interface RunCreateRequest {
  assistant_id: string;
  model?: string | null;
  instructions?: string | null;
  additional_instructions?: string | null;
  additional_messages?: { role: 'user' | 'assistant'; content: string }[];
  tools?: AssistantTool[] | null;
  metadata?: Record<string, string>;
  temperature?: number | null;
  top_p?: number | null;
  max_prompt_tokens?: number | null;
  max_completion_tokens?: number | null;
  truncation_strategy?: RunTruncationStrategy;
  tool_choice?: 'none' | 'auto' | 'required' | { type: 'function'; function: { name: string } };
  response_format?: 'auto' | { type: 'text' | 'json_object' };
}

export interface RunUpdateRequest {
  metadata?: Record<string, string>;
}

export interface RunSubmitToolOutputsRequest {
  tool_outputs: RunToolOutput[];
}

export interface RunToolOutput {
  tool_call_id: string;
  output: string;
}

export interface RunListParams {
  limit?: number;
  order?: 'asc' | 'desc';
  after?: string;
  before?: string;
}

export type RunListResponse = PaginatedResponse<Run>;

export interface RunStep {
  id: string;
  object: 'thread.run.step';
  created_at: number;
  assistant_id: string;
  thread_id: string;
  run_id: string;
  type: 'message_creation' | 'tool_calls';
  status: RunStatus;
  step_details: RunStepDetails;
  last_error: RunError | null;
  expired_at: number | null;
  cancelled_at: number | null;
  failed_at: number | null;
  completed_at: number | null;
  metadata: Record<string, string>;
  usage: RunUsage | null;
}

export type RunStepDetails = RunStepDetailsMessageCreation | RunStepDetailsToolCalls;

export interface RunStepDetailsMessageCreation {
  type: 'message_creation';
  message_creation: {
    message_id: string;
  };
}

export interface RunStepDetailsToolCalls {
  type: 'tool_calls';
  tool_calls: RunToolCall[];
}

export type RunStepListResponse = PaginatedResponse<RunStep>;

export type RunCreateParams = RunCreateRequest & RequestOptions;
export type RunUpdateParams = RunUpdateRequest & RequestOptions;
export type RunListRequestParams = RunListParams & RequestOptions;
export type RunRetrieveParams = RequestOptions;
export type RunCancelParams = RequestOptions;
export type RunSubmitToolOutputsParams = RunSubmitToolOutputsRequest & RequestOptions;
export type RunStepListParams = RunListParams & RequestOptions;
export type RunStepRetrieveParams = RequestOptions;

File: openai/rust/src/services/assistants/runs.rs
use crate::auth::AuthManager;
use crate::errors::OpenAIResult;
use crate::resilience::ResilienceOrchestrator;
use crate::transport::HttpTransport;
use crate::types::ListResponse;
use async_trait::async_trait;
use http::Method;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Arc;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Run {
    pub id: String,
    pub object: String,
    pub created_at: i64,
    pub thread_id: String,
    pub assistant_id: String,
    pub status: RunStatus,
    pub required_action: Option<serde_json::Value>,
    pub last_error: Option<RunError>,
    pub expires_at: Option<i64>,
    pub started_at: Option<i64>,
    pub cancelled_at: Option<i64>,
    pub failed_at: Option<i64>,
    pub completed_at: Option<i64>,
    pub model: String,
    pub instructions: Option<String>,
    pub tools: Vec<serde_json::Value>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<HashMap<String, String>>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RunStatus {
    Queued,
    InProgress,
    RequiresAction,
    Cancelling,
    Cancelled,
    Failed,
    Completed,
    Expired,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RunError {
    pub code: String,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RunStep {
    pub id: String,
    pub object: String,
    pub created_at: i64,
    pub assistant_id: String,
    pub thread_id: String,
    pub run_id: String,
    pub r#type: String,
    pub status: RunStatus,
    pub step_details: serde_json::Value,
}

#[async_trait]
pub trait RunService: Send + Sync {
    async fn create(&self, thread_id: &str, assistant_id: &str) -> OpenAIResult<Run>;
    async fn retrieve(&self, thread_id: &str, run_id: &str) -> OpenAIResult<Run>;
    async fn cancel(&self, thread_id: &str, run_id: &str) -> OpenAIResult<Run>;
    async fn list(&self, thread_id: &str) -> OpenAIResult<ListResponse<Run>>;
}

pub struct RunServiceImpl {
    transport: Arc<dyn HttpTransport>,
    auth_manager: Arc<dyn AuthManager>,
    resilience: Arc<dyn ResilienceOrchestrator>,
}

impl RunServiceImpl {
    pub fn new(
        transport: Arc<dyn HttpTransport>,
        auth_manager: Arc<dyn AuthManager>,
        resilience: Arc<dyn ResilienceOrchestrator>,
    ) -> Self {
        Self {
            transport,
            auth_manager,
            resilience,
        }
    }
}

#[async_trait]
impl RunService for RunServiceImpl {
    async fn create(&self, thread_id: &str, assistant_id: &str) -> OpenAIResult<Run> {
        let mut headers = http::HeaderMap::new();
        self.auth_manager.apply_auth(&mut headers).await?;
        headers.insert("OpenAI-Beta", "assistants=v2".parse().unwrap());

        let path = format!("/threads/{}/runs", thread_id);
        let body = serde_json::json!({
            "assistant_id": assistant_id,
        });

        self.resilience
            .execute(async {
                self.transport
                    .request(Method::POST, &path, Some(&body), Some(headers.clone()))
                    .await
            })
            .await
    }

    async fn retrieve(&self, thread_id: &str, run_id: &str) -> OpenAIResult<Run> {
        let mut headers = http::HeaderMap::new();
        self.auth_manager.apply_auth(&mut headers).await?;
        headers.insert("OpenAI-Beta", "assistants=v2".parse().unwrap());

        let path = format!("/threads/{}/runs/{}", thread_id, run_id);
        self.resilience
            .execute(async {
                self.transport
                    .request::<(), Run>(Method::GET, &path, None, Some(headers.clone()))
                    .await
            })
            .await
    }

    async fn cancel(&self, thread_id: &str, run_id: &str) -> OpenAIResult<Run> {
        let mut headers = http::HeaderMap::new();
        self.auth_manager.apply_auth(&mut headers).await?;
        headers.insert("OpenAI-Beta", "assistants=v2".parse().unwrap());

        let path = format!("/threads/{}/runs/{}/cancel", thread_id, run_id);
        self.resilience
            .execute(async {
                self.transport
                    .request::<(), Run>(Method::POST, &path, None, Some(headers.clone()))
                    .await
            })
            .await
    }

    async fn list(&self, thread_id: &str) -> OpenAIResult<ListResponse<Run>> {
        let mut headers = http::HeaderMap::new();
        self.auth_manager.apply_auth(&mut headers).await?;
        headers.insert("OpenAI-Beta", "assistants=v2".parse().unwrap());

        let path = format!("/threads/{}/runs", thread_id);
        self.resilience
            .execute(async {
                self.transport
                    .request::<(), ListResponse<Run>>(Method::GET, &path, None, Some(headers.clone()))
                    .await
            })
            .await
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_service_is_send_sync() {
        fn assert_send_sync<T: Send + Sync>() {}
        assert_send_sync::<RunServiceImpl>();
    }
}
